### PR TITLE
Add generic F411Rx variant

### PR DESCRIPTION
**Pull request by MCUdude**
This pinout is more or less identical to the F401Rx pinout, except for more peripherals. It follows the same pin style as the F401Rx variant.

If an external clock/oscillator is not preset, it will automatically use its own internal oscillator.

### CODE REVIEW

**Comment by MCUdude** [comment]
Seems like STm32F411RC isn't supported in the core files. Is this something that can be fixed relatively easily? 

```
Build HY_TinySTM103TB (42/95)... 
  --> Generic_F401RE SUCESS
  --> Generic_F411RC FAILED
In file included from /home/travis/.arduino15/packages/STM32/hardware/stm32/*/cores/arduino/stm32/stm32_def.h:34,
                 from /home/travis/.arduino15/packages/STM32/hardware/stm32/*/cores/arduino/stm32/clock.h:43,
                 from /home/travis/.arduino15/packages/STM32/hardware/stm32/*/cores/arduino/wiring_time.h:23,
                 from /home/travis/.arduino15/packages/STM32/hardware/stm32/*/cores/arduino/wiring.h:38,
                 from /home/travis/.arduino15/packages/STM32/hardware/stm32/*/cores/arduino/Arduino.h:32,
                 from /tmp/BuildOutput_2020-01-06_14-46-42/Generic_F411RC/sketch/BareMinimum.ino.cpp:1:
/home/travis/.arduino15/packages/STM32/hardware/stm32/*/system/Drivers/CMSIS/Device/ST/STM32F4xx/Include/stm32f4xx.h:191:3: error: #error "Please select first the target STM32F4xx device used in your application (in stm32f4xx.h file)"
  191 |  #error "Please select first the target STM32F4xx device used in your application (in stm32f4xx.h file)"
      |   ^~~~~
In file included from /home/travis/.arduino15/packages/STM32/hardware/stm32/*/cores/arduino/board.h:16,
                 from /home/travis/.arduino15/packages/STM32/hardware/stm32/*/cores/arduino/wiring.h:41,
                 from /home/travis/.arduino15/packages/STM32/hardware/stm32/*/cores/arduino/Arduino.h:32,
                 from /tmp/BuildOutput_2020-01-06_14-46-42/Generic_F411RC/sketch/BareMinimum.ino.cpp:1:
/home/travis/.arduino15/packages/STM32/hardware/stm32/*/cores/arduino/stm32/rtc.h:94:2: error: #error "Unknown Family - unknown synchronous prescaler"
   94 | #error "Unknown Family - unknown synchronous prescaler"
      |  ^~~~~
exit status 1
```

**Comment by fpistm** [comment]
> Seems like STm32F411RC isn't supported in the core files. Is this something that can be fixed relatively easily?

This simply means you didn't set the correct `product_line`:
```patch
- GenF4.menu.pnum.Generic_F411RC.build.product_line=STM32F411xC
+ GenF4.menu.pnum.Generic_F411RC.build.product_line=STM32F411xE
```

**Comment by MCUdude** [comment]
> This simply means you didn't set the correct product_line

I see. A fix is on its way! Otherwise, does it look OK?

File: variants/Generic_F411Rx/PeripheralPins.c [hunk 1]
```diff
@@ -0,0 +1,259 @@
...
+ * IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE
+ * DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER OR CONTRIBUTORS BE LIABLE
+ * FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
+ * DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR
+ * SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER
+ * CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY,
+ * OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+ * OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+ *******************************************************************************
+ * Automatically generated from STM32F411R(D-E)Tx.xml
```
Comment (fpistm) [inline variants/Generic_F411Rx/PeripheralPins.c]: Where did you found this file ?
There is no `STM32F411R(D-E)Tx.xml`, it should be `STM32F411R(C-E)Tx.xml`
https://github.com/stm32duino/Arduino_Tools/tree/master/src/genpinmap/Arduino/STM32F411R(C-E)Tx

It seems you update manually the F401 one ? Right ?
Please use the one with above link (it misses 2 I2C pins in your one).

File: variants/Generic_F411Rx/variant.h [hunk 2]
```diff
@@ -0,0 +1,149 @@
...
+#define PH0  48  // | 48      |        |           |          |                                | OSC_IN    |
+#define PH1  49  // | 49      |        |           |          |                                | OSC_OUT   |
+//                  |---------|--------|-----------|----------|--------------------------------|-----------|
+
+// This must be a literal
+#define NUM_DIGITAL_PINS        50
+#define NUM_ANALOG_INPUTS       16
+
+// On-board LED pin number
+#define LED_BUILTIN             PA5
```
Comment (fpistm) [inline variants/Generic_F411Rx/variant.h]: As it is generic you should not define it by default or at least under  #ifndef
```suggestion
#ifndef LED_BUILTIN             
#define LED_BUILTIN             PA5
#endif
```

File: boards.txt [hunk 3]
```diff
@@ -1025,6 +1025,24 @@ GenF4.menu.pnum.FEATHER_F405.build.board=FEATHER_F405
...
+GenF4.menu.pnum.Generic_F411RE.build.product_line=STM32F411xE
+GenF4.menu.pnum.Generic_F411RE.build.variant=Generic_F411Rx
+GenF4.menu.pnum.Generic_F411RE.node=NODE_F411RE
+
+# Generic F411RC
+GenF4.menu.pnum.Generic_F411RC=Generic F411RC
+GenF4.menu.pnum.Generic_F411RC.upload.maximum_size=262144
+GenF4.menu.pnum.Generic_F411RC.upload.maximum_data_size=131072
+GenF4.menu.pnum.Generic_F411RC.build.board=GENERIC_F411RC
+GenF4.menu.pnum.Generic_F411RC.build.product_line=STM32F411xC
```
[the remark this answers is not in the document]
Comment (fpistm) [reply boards.txt]: OK this one is fixed.

File: variants/Generic_F411Rx/PeripheralPins.c (hunk 1 above)
Comment (MCUdude) [reply variants/Generic_F411Rx/PeripheralPins.c]: Yes, I used the F401 as base. But I can't copy the contents from the PeripheralPins.c you linked in directly? Some lines have to be commented out like done on other pinouts?

Comment (fpistm) [reply variants/Generic_F411Rx/PeripheralPins.c]: Yes you should comments some lines, it is user choice to select his peripheral pins mappings and this is the reason why all arrays are WEAKed to be able to redefine them as you want.
As the first pin found in the selected array is used then having multiple one is useless, so comment them out allow to save space.

**Comment by MCUdude** [comment]
Should be OK now. Now everything is only one commit too.

File: variants/Generic_F411Rx/PeripheralPins.c [hunk 4]
```diff
@@ -0,0 +1,271 @@
...
+WEAK const PinMap PinMap_PWM[] = {
+  //  {PA_0,  TIM2,   STM_PIN_DATA_EXT(STM_MODE_AF_PP, GPIO_PULLUP, GPIO_AF1_TIM2, 1, 0)}, // TIM2_CH1
+  {PA_0,  TIM5,   STM_PIN_DATA_EXT(STM_MODE_AF_PP, GPIO_PULLUP, GPIO_AF2_TIM5, 1, 0)}, // TIM5_CH1
+  //  {PA_1,  TIM2,   STM_PIN_DATA_EXT(STM_MODE_AF_PP, GPIO_PULLUP, GPIO_AF1_TIM2, 2, 0)}, // TIM2_CH2
+  {PA_1,  TIM5,   STM_PIN_DATA_EXT(STM_MODE_AF_PP, GPIO_PULLUP, GPIO_AF2_TIM5, 2, 0)}, // TIM5_CH2
+  //  {PA_2,  TIM2,   STM_PIN_DATA_EXT(STM_MODE_AF_PP, GPIO_PULLUP, GPIO_AF1_TIM2, 3, 0)}, // TIM2_CH3
+  {PA_2,  TIM5,   STM_PIN_DATA_EXT(STM_MODE_AF_PP, GPIO_PULLUP, GPIO_AF2_TIM5, 3, 0)}, // TIM5_CH3
+  //  {PA_2,  TIM9,   STM_PIN_DATA_EXT(STM_MODE_AF_PP, GPIO_PULLUP, GPIO_AF3_TIM9, 1, 0)}, // TIM9_CH1
+  //  {PA_3,  TIM2,   STM_PIN_DATA_EXT(STM_MODE_AF_PP, GPIO_PULLUP, GPIO_AF1_TIM2, 4, 0)}, // TIM2_CH4
+    {PA_3,  TIM5,   STM_PIN_DATA_EXT(STM_MODE_AF_PP, GPIO_PULLUP, GPIO_AF2_TIM5, 4, 0)}, // TIM5_CH4
```
Comment (fpistm) [inline variants/Generic_F411Rx/PeripheralPins.c]:
```suggestion
  {PA_3,  TIM5,   STM_PIN_DATA_EXT(STM_MODE_AF_PP, GPIO_PULLUP, GPIO_AF2_TIM5, 4, 0)}, // TIM5_CH4
```

**Comment by MCUdude** [comment]
I'm sorry, should be fixed now.

**Comment by fpistm** [comment]
I saw you last change, I do not check those comments, the reference is the PeripheralPins.c file which reference all the possible value, that's why it is important to have it up to date and for the dedicated MCU version.

**Comment by MCUdude** [comment]
I was looking through some source files for the STM32F411Rx series when I discovered that I had done a mistake when creating the table in variants.h. The contents of PeripheralPins.c is OK. I just find it annoying when comments like this don't match the actual source.

**Comment by MCUdude** [comment]
I see this is left open. Anything else that should be changed/fixed?

**Pull request merged**